### ISSUE TITLE
Error when running quint verify: EXCEPT

### ISSUE THREAD

**Issue by p-offtermatt** [informalsystems/quint#1408]
Hey, another error from my recent tries to get my model to work with Apalache.
Spec: https://github.com/cosmos/interchain-security/blob/6ec927d0005be644e818f71eefa8f9446462bcaf/tests/mbt/model/ccv.qnt#L939

```
% quint verify ccv_model.qnt --verbosity 5
'Input was not a valid representation of the QuintIR: Conversion failed while building operator definition `ccv_model::ccv::recvVscMaturedPacketOnProvider`: Operator EXCEPT cannot be applied to arguments of types ({ acksToSendOnEndBlock: (Str -> Seq(Slash({ downtime: Bool, id: Int, sendingTime: Int, timeoutTime: Int, valPower: Int, validator: Str }) | Vsc({ downtimeSlashAcks: Seq(Str), id: Int, sendingTime: Int, timeoutTime: Int, validatorSet: (Str -> Int) }) | VscMatured({ id: Int, sendingTime: Int, timeoutTime: Int }))), chainState: { currentBlockHeight: Int, currentValidatorPowers: (Str -> Int), jailedUntil: (Str -> Int), lastTimestamp: Int, runningTimestamp: Int, votingPowerHistory: Seq((Str -> Int)) }, consumerAddrToValidator: (Str -> (a47 -> a48)), consumerAddrsToPrune: (Str -> (Int -> Seq(a49))), consumerStatus: (Str -> Str), consumersWithAddrAssignmentChangesInThisEpoch: Set(Str), consumersWithPowerChangesInThisEpoch: Set(Str), downtimeSlashRequests: (Str -> Seq(Str)), keyAssignedValSetHistory: (Str -> Seq((Str -> Int))), keyAssignmentsForVSCPackets: (Int -> (Str -> (Str -> Str))), outstandingPacketsToConsumer: (Str -> Seq(Slash({ downtime: Bool, id: Int, sendingTime: Int, timeoutTime: Int, valPower: Int, validator: Str }) | Vsc({ downtimeSlashAcks: Seq(Str), id: Int, sendingTime: Int, timeoutTime: Int, validatorSet: (Str -> Int) }) | VscMatured({ id: Int, sendingTime: Int, timeoutTime: Int }))), receivedMaturations: Set({ id: Int, sendingTime: Int, timeoutTime: Int }), runningVscId: Int, sentVscPacketsToConsumer: (Str -> Seq({ downtimeSlashAcks: Seq(Str), id: Int, sendingTime: Int, timeoutTime: Int, validatorSet: (Str -> Int) })), validatorToConsumerAddr: (Str -> (Str -> Str)) }, <<Str>>, (Str -> (a54 -> a55)))'
```

**Comment by p-offtermatt** [comment]
For reference, here is the function it flags:
```
pure def recvVscMaturedPacketOnProvider(currentState: ProtocolState, sender: Chain, packet: VscMaturedPacket): Result = {
        if (currentState.providerState.sentVscPacketsToConsumer.get(sender).head().id != packet.id) {
            // the packet is not the oldest sentVscPacket, something went wrong
            Err("Received maturation is not for the oldest sentVscPacket")
        } else {
            // add the packet to the list of received maturations.
            val currentReceivedMaturations = currentState.providerState.receivedMaturations
            val newReceivedMaturations = currentReceivedMaturations.union(Set(packet))
            val newProviderState = currentState.providerState.with(
                "receivedMaturations", newReceivedMaturations
            )
            // prune the sentVscPacket
            val newSentVscPacket = currentState.providerState.sentVscPacketsToConsumer.get(sender).tail()
            val newState = currentState.with(
                "providerState", 
                    {...newProviderState,
                    sentVscPacketsToConsumer: currentState.providerState.sentVscPacketsToConsumer.set(sender, newSentVscPacket)
                    }
            )

            // prune assigned keys that are no longer needed

            // get the keys that should be pruned
            pure val consumerAddrsToPrune = 
                currentState.providerState.consumerAddrsToPrune
                    .getOrElse(sender, Map())
                    .getOrElse(packet.id, List()).toSet()

            // actually do the pruning
            pure val senderValByConsAddr = 
                // for the sender chain, get the consAddrToValidator mapping
                newState.providerState.consumerAddrToValidator.getOrElse(sender, Map())
            pure val newSenderValByConsAddr =
                senderValByConsAddr.keys().filter(
                    // filter out the assigned keys that should be pruned
                    consAddr => not(consumerAddrsToPrune.contains(consAddr))
                ).mapBy(
                    // rebuild the map with the pruned keys removed
                    consAddr => senderValByConsAddr.get(consAddr)
                )

            // update the provider state
            pure val newProviderState2 = newState.providerState.with(
                // update the consumerAddrToValidator map to remove the pruned keys
                "consumerAddrToValidator", 
                newState.providerState.consumerAddrToValidator.put(sender, newSenderValByConsAddr)
            ).with(
                // delete the consumer addresses to prune, since they are pruned now
                "consumerAddrsToPrune", 
                newState.providerState.consumerAddrsToPrune.put(
                    sender, 
                    newState.providerState.consumerAddrsToPrune.getOrElse(sender, Map()).put(
                        packet.id, 
                        List()
                    )
                )
            )
            // update the state
            pure val newState2 = newState.with(
                "providerState", newProviderState2
            )

            Ok(newState2)
        }
```

(if I comment out the function body and replace it with a trivial return, I don't get this error anymore, so I assume it is somewhere in here)

**Comment by bugarela** [comment]
I can't reproduce this error anymore, I assume my type checker fix from #1409 also fixed this :tada:
